Add static pages management and dynamic routes

aabdaab opened the pull request:
This PR adds models for static pages (think the content of `About` tab):
* Page - that will hold its title, slug and short description
* Content - content of the page, page can have many contents - mostly did this because of `Supporters` page that has two different sections with related logos. Content will have text and images.
* DynamicRouter - for adding dynamic routes when new page is added.

I added the `Pages` section in ActiveAdmin, under TPI:
<img width="943" alt="Screenshot 2019-11-25 at 13 30 16" src="https://user-images.githubusercontent.com/6136899/69544495-c0ed4c00-0f87-11ea-8c7a-82a714357739.png">

I tried to follow the example of the `Litigations` form that has associations with `litigation_sides` - I wanted to do an equivalent for Page and its Contents.

When you add a new Page, you can go to `/tpi/:slug` to see that the route was created.

Before testing:
- run `rake db:migrate`

tsubik commented on db/schema.rb:
```diff
@@ -136,6 +145,11 @@
     t.index ["documentable_type", "documentable_id"], name: "index_documents_on_documentable_type_and_documentable_id"
   end
 
+  create_table "dynamic_routers", force: :cascade do |t|
```
Why do we need this table?

tsubik commented on app/models/dynamic_router.rb:
```diff
@@ -0,0 +1,14 @@
+class DynamicRouter < ApplicationRecord
```
That could be a service?

tsubik replied:
Or we need this to be persisted model in DB?

aabdaab replied:
I guess it could, it'd be more verbose.. let me check

tsubik replied:
I mean it could be just moved to `services` and not inherit after `ApplicationRecord` if we don't need the model and the table

aabdaab commented:
Fixed saving the contents ☝️Also noticed that tests are not failing

tsubik commented on app/models/page.rb:
```diff
@@ -0,0 +1,16 @@
+    accepts_nested_attributes_for :contents
+  end
+
+  after_save :reload_routes
```
I would do this in `after_commit`

tsubik commented on app/admin/page.rb:
```diff
@@ -3,7 +3,7 @@
   menu priority: 6, parent: 'TPI'
 
-  permit_params :title, :slug, :description, contents_attributes: permit_params_for(:contents), content_ids: []
+  permit_params :title, :slug, :description, :active_tab, contents_attributes: [:title, :text], content_ids: []
```
I don't think we need :active_tab here

aabdaab commented:
Addressed the comments ☝️ 
I'm checking now why the tests are failing

aabdaab commented:
* Fixed tests
* Added uploading images
* Fixed issue with duplicated content records on the edit form - due to lack of `:id` in permitted params for content

There's still an issue with rendering the already uploaded images (`image_tag` apparently not working with ActiveStorage this way) and uploading more images in the edit form once we already uploaded some - but I think it can be handled in the next PRs?

tsubik commented:
Great. I can take a look at those images. Checking this now.

tsubik commented on app/helpers/contents_helper.rb:
```diff
@@ -0,0 +1,2 @@
+module ContentsHelper
```
we can remove that

tsubik commented on app/helpers/pages_helper.rb:
```diff
@@ -0,0 +1,2 @@
+module PagesHelper
```
this too

aabdaab replied:
👍

tsubik commented on spec/factories/dynamic_routers.rb:
```diff
@@ -0,0 +1,4 @@
+FactoryBot.define do
```
this too, to remove

tsubik commented on spec/decorators/content_decorator_spec.rb:
```diff
@@ -0,0 +1,4 @@
+require 'rails_helper'
```
hmm, not sure if we need to keep empty tests

tsubik commented on app/assets/stylesheets/contents.scss:
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Contents controller here.
```
to remove

tsubik commented on app/controllers/tpi/contents_controller.rb:
```diff
@@ -0,0 +1,4 @@
+module TPI
```
we dont need that too.

simaob commented on config/routes.rb:
```diff
@@ -15,7 +15,7 @@
       end
     end
-    
+
```
just trailing spaces removal!

tsubik commented:
@aabdaab you could use decorators and for all html content use `html_safe` take a look at `TargetDecorator` for example.

aabdaab commented:
I added a few changes to this PR:
* new `Image` model that holds a `logo` and a `link`
* added a rake task `rake static_pages:generate` that's gonna scaffold some known Pages and its Contents - you can then change it in Active Admin, add logos to it, links etc.

![image](https://user-images.githubusercontent.com/6136899/69643537-5efe1680-105b-11ea-829a-c131ded49371.png)

aabdaab commented:
I added a rake task that creates some of the `Page` records beforehand and changed the `DynamicRouter`'s condition to pick up the routes and now the tests are failing again (on Travis, locally they're passing). I'm gonna look into this in more depth tomorrow.
